cpu+board: Add support for NXP Kinetis KW41Z and FRDM-KW41Z development board

gebart opened the pull request:
Port RIOT to the recently released NXP FRDM-KW41Z development boards.

Currently missing support for ~~the radio and~~ the onboard NOR flash.
I2C needs a rewrite, this CPU has one i2c module clocked from the bus clock, and one clocked by the system clock. The I2C rewrite will be on hold until the I2C API is reworked, so for now, one of the two available I2C buses will be clocked at the wrong speed, this can be worked around in the application by requesting a different speed or modifying the configuration to fit.

A radio transceiver driver is available in #7107  

depends on ~~#7362~~ ~~#6994~~ ~~#6978~~ ~~#6916~~ ~~#6993~~ ~~#7434~~ ~~#7882~~ ~~#8560~~

gebart commented:
This requires a patched openocd which can be found at https://github.com/gebart/openocd/tree/pr/kwxxz-support?files=1

gebart commented:
The required openocd patches have been merged into the upstream master branch for what I assume will be released as openocd version 0.11.

emmanuelsearch commented:
@gebart meaning that I could just clone master from http://git.code.sf.net/p/openocd/code and use that?

gebart commented:
Yes, I believe so. I did not test it myself yet, but I got some emails yesterday from the Gerrit system that my patches have been merged. If openocd doesn't have the new CPU support you will get a message saying "unsupported MCU" or something similar and the flashing will fail. 

I think you may need to have libusb-dev installed for building openocd with jlink support. The configure script will enable all supported interfaces automatically based on which libraries are installed on your machine and you will see a summary of which interfaces are enabled at the end of configure, if jlink is not enabled you will need some other lib.

kYc0o commented:
I have a board now so I can test and review the code. I'll try to merge this on today's Hack'n'ACK

kYc0o commented:
Well, I think I have everything configured now but no success while flashing:

```console
### Flashing Target ###
/Users/facosta/git/RIOT2/RIOT/examples/default/bin/frdm-kw41z/default.elf is fine.
Open On-Chip Debugger 0.10.0+dev-00161-g1725abc (2017-06-27-19:26)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : add flash_bank kinetis klx.pflash
adapter speed: 1000 kHz
none separate
cortex_m reset_config sysresetreq
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : No device selected, using first device.
Info : J-Link OpenSDA 2 compiled May  6 2016 11:04:17
Info : Hardware version: 1.00
Info : VTarget = 3.300 V
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x0bc11477
Error: MDM: failed to read ID register
Info : SWD DPIDR 0x0bc11477
Info : SWD DPIDR 0x0bc11477
Error: Failed to read memory at 0xe000ed00
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* klx.cpu            cortex_m   little klx.cpu            reset
Polling target klx.cpu failed, trying to reexamine
Info : SWD DPIDR 0x0bc11477
Warn : **** Your Kinetis MCU is probably locked-up in RESET/WDOG loop. ****
Warn : **** Common reason is a blank flash (at least a reset vector).  ****
Warn : **** Issue 'kinetis mdm halt' command or if SRST is connected   ****
Warn : **** and configured, use 'reset halt'                           ****
Warn : **** If MCU cannot be halted, it is likely secured and running  ****
Warn : **** in RESET/WDOG loop. Issue 'kinetis mdm mass_erase'         ****
Info : SWD DPIDR 0x0bc11477
Error: Failed to read memory at 0xe000edfc
Info : SWD DPIDR 0x0bc11477
Error: Failed to read memory at 0xe0001000
Info : klx.cpu: hardware has 0 breakpoints, 0 watchpoints
Warn : **** Your Kinetis MCU is probably locked-up in RESET/WDOG loop. ****
Warn : **** Common reason is a blank flash (at least a reset vector).  ****
Warn : **** Issue 'kinetis mdm halt' command or if SRST is connected   ****
Warn : **** and configured, use 'reset halt'                           ****
Warn : **** If MCU cannot be halted, it is likely secured and running  ****
Warn : **** in RESET/WDOG loop. Issue 'kinetis mdm mass_erase'         ****
target halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0xfffffffe msp: 0xfffffffc
auto erase enabled
Info : Kinetis MKW41Z512xxx4 detected: 2 flash blocks
Info : 2 PFlash banks: 512k total
Info : This device supports Program Longword execution only.
Info : Disabling Kinetis watchdog (initial SIM_COPC 0x0c)
Info : This device supports Program Longword execution only.
Error: jaylink_swd_io() failed: JAYLINK_ERR_DEV_NO_MEMORY.
Warn : Sending data to device timed out, retrying.
Warn : Sending data to device timed out, retrying.
Error: Sending data to device timed out.
Error: transport_write() failed: timeout occurred.
Error: jaylink_swd_io() failed: JAYLINK_ERR_TIMEOUT.
Warn : Last read operation left 16 bytes.
Warn : Sending data to device timed out, retrying.
Error: Failed to send data to device: input/output error.
Error: transport_write() failed: input/output error.
Error: jaylink_swd_io() failed: JAYLINK_ERR_IO.
Warn : Last read operation left 16 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Error: klx.cpu -- clearing lockup after double fault
Warn : Last read operation left 34 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 9 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 15 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 9 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 33 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 9 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 33 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 22 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 16 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 33 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Error: klx.cpu -- clearing lockup after double fault
Warn : Last read operation left 33 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Warn : Last read operation left 9 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Error: error waiting for target flash write algorithm
Error: Error executing kinetis Flash programming algorithm
Warn : Last read operation left 15 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_swd_io() failed: JAYLINK_ERR.
Error: error writing to flash at address 0x00000000 at offset 0x00000000

Warn : Last read operation left 28 bytes.
Error: Failed to send data to device: LIBUSB_ERROR_PIPE.
Error: transport_write() failed: unspecified error.
Error: jaylink_unregister() failed: JAYLINK_ERR.
make: *** [flash] Error 1
```

Maybe I need to press some buttons or something?

gebart commented:
@kYc0o try running make debug-server and connect with telnet to openocd and run the listed command (kinetis mdm mass_erase) followed by reset halt. Then you can kill the debug server and try make flash again

kYc0o commented:
I have tried that without success (same error as above)... I'll try today to flash it using a Segger J-Link adapter, maybe there's still some things to erase or program before having a functional openocd programming.

However, it's quite uncomfortable this way, since it means we cannot flash these devices out of the box. You board was flashable without any problem since the beginning? Have you updated the J-Link firmware? It's maybe only my board which is concerned...

Once I succeed, maybe we can merge this as a first or "early" support, since anyways we need to compile the latest master of openocd to have support for flashing this board.

gebart commented:
Now that you mention it, my board was factory flashed with some LED blink app, so it wasn't stuck at all. I have only used openocd together with the on board jlink adapter. I did reflash one of my boards with a cmsis dap firmware just to try it out, and that one works as well (with FRDM_IFACE=dap)

kYc0o commented:
Using an external J-Link adapter the flashing is successful, then if I try again with the default (openocd with onboard jlink) I still have the errors.

What I notice is that the flasher doesn't event succeed to read the MDM number, which I think is essential. However, the external J-Link succeeds even with this errors.

Below the output of a successful flashing with an external Segger J-Link:

```console
Open On-Chip Debugger 0.10.0+dev-00161-g1725abc (2017-06-27-20:01)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : add flash_bank kinetis klx.pflash
adapter speed: 1000 kHz
none separate
cortex_m reset_config sysresetreq
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : No device selected, using first device.
Info : J-Link V9 compiled Feb  2 2016 18:43:46
Info : Hardware version: 9.30
Info : VTarget = 2.985 V
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x0bc11477
Error: MDM: failed to read ID register
Info : SWD DPIDR 0x0bc11477
Info : SWD DPIDR 0x0bc11477
Error: Failed to read memory at 0xe000edfc
Info : klx.cpu: hardware has 0 breakpoints, 2 watchpoints
Polling target klx.cpu failed, trying to reexamine
Error: The 'check_security' command must be used after 'init'.
in procedure 'init' 
in procedure 'ocd_bouncer' 
in procedure 'flash' 
in procedure 'ocd_bouncer'

Info : SWD DPIDR 0x0bc11477
Info : SWD DPIDR 0x0bc11477
Error: Failed to read memory at 0xe000edfc
Info : SWD DPIDR 0x0bc11477
Error: Failed to write memory at 0xe0001028
Info : klx.cpu: hardware has 0 breakpoints, 2 watchpoints
Polling target klx.cpu failed, trying to reexamine
Info : SWD DPIDR 0x0bc11477
Warn : **** Your Kinetis MCU is probably locked-up in RESET/WDOG loop. ****
Warn : **** Common reason is a blank flash (at least a reset vector).  ****
Warn : **** Issue 'kinetis mdm halt' command or if SRST is connected   ****
Warn : **** and configured, use 'reset halt'                           ****
Warn : **** If MCU cannot be halted, it is likely secured and running  ****
Warn : **** in RESET/WDOG loop. Issue 'kinetis mdm mass_erase'         ****
Info : SWD DPIDR 0x0bc11477
Error: Failed to read memory at 0xe000ed00
Examination failed, GDB will be halted. Polling again in 100ms
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* klx.cpu            cortex_m   little klx.cpu            unknown
Warn : **** Your Kinetis MCU is probably locked-up in RESET/WDOG loop. ****
Warn : **** Common reason is a blank flash (at least a reset vector).  ****
Warn : **** Issue 'kinetis mdm halt' command or if SRST is connected   ****
Warn : **** and configured, use 'reset halt'                           ****
Warn : **** If MCU cannot be halted, it is likely secured and running  ****
Warn : **** in RESET/WDOG loop. Issue 'kinetis mdm mass_erase'         ****
target halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0xfffffffe msp: 0xfffffffc
auto erase enabled
Info : Kinetis MKW41Z512xxx4 detected: 2 flash blocks
Info : 2 PFlash banks: 512k total
Info : This device supports Program Longword execution only.
Info : Disabling Kinetis watchdog (initial SIM_COPC 0x0c)
Info : This device supports Program Longword execution only.
Warn : Flash Configuration Field written.
Warn : Reset or power off the device to make settings effective.
Info : This device supports Program Longword execution only.
wrote 12288 bytes from file /Users/facosta/git/RIOT2/RIOT/examples/hello-world/bin/frdm-kw41z/hello-world.elf in 0.422928s (28.374 KiB/s)
Info : MDM: Chip is unsecured. Continuing.
target halted due to debug-request, current mode: Thread 
xPSR: 0x61000000 pc: 0x00000764 msp: 0x1fff8200
verified 11472 bytes in 0.139199s (80.483 KiB/s)
Info : MDM: Chip is unsecured. Continuing.
shutdown command invoked
Done flashing
```

gebart commented:
Did you try reflashing  the debugger on the board? 
Either the latest jlink frdm-kw41z firmware from Segger, or the mbed cmsis dap from nxp.com

Also, do you have the same problem with both boards? (I assume you have two since the frdm-kw41z  box comes with two identical boards)

kYc0o commented:
I was just doing that when you posted your comment. Now I have different errors:

```console
### Flashing Target ###
/Users/facosta/git/RIOT2/RIOT/examples/hello-world/bin/frdm-kw41z/hello-world.elf is fine.
Open On-Chip Debugger 0.10.0+dev-00161-g1725abc (2017-06-27-20:01)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : add flash_bank kinetis klx.pflash
adapter speed: 1000 kHz
none separate
cortex_m reset_config sysresetreq
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : No device selected, using first device.
Info : J-Link OpenSDA 2 compiled May  6 2016 11:04:17
Info : Hardware version: 1.00
Info : VTarget = 3.300 V
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x0bc11477
Info : MDM: Chip is unsecured. Continuing.
Info : klx.cpu: hardware has 2 breakpoints, 2 watchpoints
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* klx.cpu            cortex_m   little klx.cpu            halted
Info : MDM: Chip is unsecured. Continuing.
target halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x0001541c msp: 0x20000428
auto erase enabled
Info : Kinetis MKW41Z512xxx4 detected: 2 flash blocks
Info : 2 PFlash banks: 512k total
Error: Flash operation failed, protection violated
Warn : erase sector 0 failed
Error: failed erasing sectors 0 to 5

make: *** [flash] Error 1
```

I don't know if I succeeded by updating it since I still have `Hardware version: 1.00`.

I can try with CMSIS-DAP and post the output.

gebart commented:
@kYc0o did you move the jumper near the reset button to connect the reset button to the debugger instead of to the target? You need to move the jumper from the default position. 
See https://community.nxp.com/docs/DOC-333763

Hold the reset button while connecting the USB cable, and keep holding it until the board shows up on the PC. 
If it works, the USB mass storage device will be labeled BOOT or BOOTLOADER, instead of FRDM-KW41Z or DAPLINK. 
Copy the debugger firmware image to the bootloader USB drive, the board will reset and reconnect after a few seconds.

kYc0o commented:
Thanks! I succeeded to update the firmware but no succeess so far... I uploaded a CMSIS-DAP firmware but it's not recognised by default. Do I need to set some variable like "PROGRAMMER" or something like that?

Regarding the second board, unfortunately I only grabbed one from @emmanuelsearch, who still has it afaik.

kYc0o commented:
I modified the interface to `dap` instead of `jlink`, which seems to recognise the dap interface but still no flashing...

```console
### Flashing Target ###
/Users/facosta/git/RIOT2/RIOT/examples/hello-world/bin/frdm-kw41z/hello-world.elf is fine.
Open On-Chip Debugger 0.10.0+dev-00161-g1725abc (2017-06-27-20:01)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : add flash_bank kinetis klx.pflash
adapter speed: 1000 kHz
none separate
cortex_m reset_config sysresetreq
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : CMSIS-DAP: FW Version = 1.0
Info : SWCLK/TCK = 0 SWDIO/TMS = 1 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x0bc11477
Error: MDM: failed to read ID register
Info : SWD DPIDR 0x0bc11477
Info : SWD DPIDR 0x0bc11477
Error: Failed to write memory at 0xe0002008
Info : SWD DPIDR 0x0bc11477
Error: Failed to read memory at 0xe0001000
Info : klx.cpu: hardware has 2 breakpoints, 0 watchpoints
Polling target klx.cpu failed, trying to reexamine
Error: The 'check_security' command must be used after 'init'.
in procedure 'init' 
in procedure 'ocd_bouncer' 
in procedure 'flash' 
in procedure 'ocd_bouncer'

Info : SWD DPIDR 0x0bc11477
Info : SWD DPIDR 0x0bc11477
Error: Failed to write memory at 0xe000200c
Info : klx.cpu: hardware has 2 breakpoints, 2 watchpoints
Polling target klx.cpu failed, trying to reexamine
Error: The 'check_security' command must be used after 'init'.
in procedure 'init' 
in procedure 'ocd_bouncer' 
in procedure 'pld' 
in procedure 'ocd_bouncer'

Info : SWD DPIDR 0x0bc11477
Info : SWD DPIDR 0x0bc11477
Error: Failed to read memory at 0xe0001000
Info : klx.cpu: hardware has 2 breakpoints, 0 watchpoints
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* klx.cpu            cortex_m   little klx.cpu            reset
Warn : **** Your Kinetis MCU is probably locked-up in RESET/WDOG loop. ****
Warn : **** Common reason is a blank flash (at least a reset vector).  ****
Warn : **** Issue 'kinetis mdm halt' command or if SRST is connected   ****
Warn : **** and configured, use 'reset halt'                           ****
Warn : **** If MCU cannot be halted, it is likely secured and running  ****
Warn : **** in RESET/WDOG loop. Issue 'kinetis mdm mass_erase'         ****
target halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x0001541c msp: 0x20000428
auto erase enabled
Info : Kinetis MKW41Z512xxx4 detected: 2 flash blocks
Info : 2 PFlash banks: 512k total
Error: Flash operation failed, protection violated
Warn : erase sector 0 failed
Error: failed erasing sectors 0 to 5

make: *** [flash] Error 1
```

kYc0o commented:
And now I think I bricked my device... no flashing possible even with the external Segger J-Link...

gebart commented:
Btw, the environment variable for selecting jlink or dap is named FRDM_IFACE. 
`export FRDM_IFACE=dap`

kYc0o commented:
With an external power source I succeeded to flash with the external Segger J-Link, then I tried again with the dap interface and worked too. Then I reflashed OpenSDA and it does't work again.

I don't know what's my problem but I'll leave CMSIS-DAP for the next tests.

kYc0o commented on cpu/kw41z/vectors.c:
```diff
@@ -0,0 +1,291 @@
+    dummy_handler,         /* reserved 252 */
+    dummy_handler,         /* reserved 253 */
+    dummy_handler,         /* reserved 254 */
+    dummy_handler,         /* reserved 255 */
```
Are all this dummy handlers really necessary?

gebart replied:
There is a check in the linker script to ensure that the interrupt vector and flash security blocks are placed at the correct location and size.

kYc0o replied:
Ok great, I understand now.

kYc0o commented:
I just tested the saul options with the default example and none seem to work. Is it expected?

gebart commented on boards/frdm-kw41z/include/board.h:
```diff
@@ -0,0 +1,101 @@
+#define XTIMER_HZ                   (1000000ul)
+#define XTIMER_SHIFT                (0)
+#endif
+/** @} */
```
The current configuration is to run xtimer with a 32768 Hz timer, to allow low power modes. The disabled xtimer configuration is for a 1 MHz timer, but that disallows low power mode.

kYc0o replied:
Ok perfect, thanks for the explanation.

gebart commented on boards/frdm-kw41z/include/board.h:
```diff
@@ -0,0 +1,101 @@
+
+#define LED0_ON            (GPIOC->PCOR = LED0_MASK)
+#define LED0_OFF           (GPIOC->PSOR = LED0_MASK)
+#define LED0_TOGGLE        (GPIOC->PTOR = LED0_MASK)
```
I did not look into it, but it should not be difficult.

kYc0o replied:
Ok well, this is a minor comment, can come in another PR if you want.

gebart replied:
the board design is a bit conflicted, the pins used for the LEDs are also connected to the headers. I will post an updated version tomorrow

gebart replied:
added the remaining colors below

gebart commented on boards/frdm-kw41z/Makefile.dep:
```diff
@@ -0,0 +1,7 @@
+# USEMODULE += mtd_spi_nor
+
+ifneq (,$(filter saul_default,$(USEMODULE)))
+#   USEMODULE += saul_gpio
```
Missing configuration file

kYc0o replied:
Well the thing is that SAUL seems to "work" or at least active on this board, but giving weird numbers. Is it possible to disable SAUL on this board until we have the configuration files?

kYc0o commented:
@gebart thanks for the explanations! Will you have time to address the rest of the comments so we can merge this? Thanks a lot for the efforts!

gebart commented:
We need to merge #6994 first since this PR depends on it. I think I will have some time to address the comments later during this week.

emmanuelsearch commented:
Finally had the time to upgrade my OpenOCD. I was able to build, flash and run #6994 #6995 and #7107 combined from Vagrant on my KW41Z, and I could ping a SAMR21 using the gnrc_networking example.

gebart commented:
@emmanuelsearch thanks for testing the RF interoperability with another board!

gebart commented:
Rebased

gebart commented on boards/frdm-common/Makefile.include:
```diff
@@ -0,0 +1,46 @@
+# set default port depending on operating system
+PORT_LINUX ?= /dev/ttyACM0
+
```
addressed in #6994 before it was merged

gebart commented on boards/frdm-kw41z/include/board.h:
```diff
@@ -0,0 +1,101 @@
+ */
+/** @{ */
+/* extern mtd_dev_t *mtd0; */
+#define MTD_0 mtd0
```
The reason for the disabled MTD driver is that the onboard SPI NOR uses a different command set than the default mtd_spi_nor command set. I don't understand why the designers chose that chip when it is different from the majority of flash chips out there...
I'll remove the disabled code and leave it open for anyone to contribute. I don't think I will have time to implement the different MTD command set myself.

gebart commented:
Rebased

gebart commented:
I will look into the SAUL configuration for this board

gebart commented:
- rebased
 - addressed comments
 - improved the SAUL configuration
 - tested all SAUL signals and they seem to be working
 - removed the MTD references

gebart commented on boards/frdm-kw41z/board.c:
```diff
@@ -79,8 +37,11 @@ void board_init(void)
+    gpio_init(LED1_PIN, GPIO_OUT);
+    gpio_set(LED1_PIN);
+    gpio_init(LED2_PIN, GPIO_OUT);
+    gpio_set(LED2_PIN);
```
SAUL turns these on again when saul is initialized, this is for applications without SAUL

gebart commented:
Removed the WIP flag, I think this is ready for review.

haukepetersen commented on boards/frdm-kw41z/Makefile.features:
```diff
@@ -0,0 +1,18 @@
+FEATURES_PROVIDED += periph_gpio
+#FEATURES_PROVIDED += periph_hwrng
+FEATURES_PROVIDED += periph_i2c
+#FEATURES_PROVIDED += periph_pwm
```
Simply remove this line if no PWM is configured...

haukepetersen commented on boards/frdm-kw41z/include/board.h:
```diff
@@ -0,0 +1,109 @@
+#define XTIMER_ISR_BACKOFF          (40)
+#define XTIMER_OVERHEAD             (30)
+#define XTIMER_HZ                   (1000000ul)
+#define XTIMER_SHIFT                (0)
```
Some of these values are the default values, right (e.g. XTIMER_HZ, SHIFT, ...)? I would prefer to only include the values here that are not default.

haukepetersen commented on boards/frdm-kw41z/include/board.h:
```diff
@@ -0,0 +1,109 @@
+ */
+#if 1
+/* LPTMR xtimer configuration */
+/* WIP, Use PIT for now */
```
this is not true currently, is it?

It could also be nice, if you would add a small comment, that per default we use the rtt as base for the xtimer, but that one can simply switch to the PIT timer by changing to `#if 0`. Or better: how about we guard this block with something like 
```c
#ifndef FRDM_XTIMER_SOURCE_PIT
...
#else
...
#endif
```
This way one could switch the xtimer configuration via env vars or in the APP makefile

haukepetersen commented on boards/frdm-kw41z/include/board.h:
```diff
@@ -0,0 +1,101 @@
+ */
+/** @{ */
+/* extern mtd_dev_t *mtd0; */
+#define MTD_0 mtd0
```
again: copy your comment above into the doxygen :-)

haukepetersen commented on boards/frdm-kw41z/include/periph_conf.h:
```diff
@@ -0,0 +1,300 @@
+ * @name    PWM configuration
+ * @{
+ */
+#define PWM_NUMOF           (0)
```
This block can/should be removed

gebart commented:
@haukepetersen addressed your comments.

TODO, separate PR: Move Kinetis xtimer configuration to `kinetis_common/include/something`

gebart commented:
@kYc0o are you satisfied with the answers to your comments, and the added changes?

gebart commented:
ping @kYc0o @haukepetersen 
Can we get this board merged for the 2018.01 release?

haukepetersen commented:
> Can we get this board merged for the 2018.01 release?

Yes, definitely!

haukepetersen commented:
just need to find time for another good review :-)

haukepetersen commented:
Also #7362 is merged, so any rebase needed?

gebart commented:
@haukepetersen I rebased this on top of #7882 and #7897, because it was easier than doing the reverse, so this currently depends on those PRs now.

emmanuelsearch commented:
@gebart #7882 is merged.
Does it makes sense to *not* make this PR depend on low-power mode work, and add this separately, later?
We could then quickly (?) merge this and move on?

gebart commented:
@emmanuelsearch I guess so, the pm PR and discussion could drag on for a while. I will try to disentangle this from the pm stuff.

gebart commented:
I am going to need to squash this to make the rebase simpler.

gebart commented:
rebased and all comments have been addressed, please check again
ping @emmanuelsearch @kYc0o @haukepetersen

gebart commented:
Updated the SAUL configuration to invert the LED pins similar to #8067 

This is ready for review

haukepetersen commented on boards/frdm-common/dist/openocd-dap-klx.cfg:
```diff
@@ -0,0 +1,25 @@
+#
```
Just wondering, could this be adapted to #7686?

gebart replied:
Yes, will update

gebart replied:
addressed

haukepetersen commented on boards/frdm-kw41z/board.c:
```diff
@@ -0,0 +1,51 @@
+static inline void set_lpuart_clock_source(void)
+{
+    /* Use MCGIRCLK (internal reference 4 MHz clock) */
+    SIM->SOPT2 = (SIM->SOPT2 & ~SIM_SOPT2_LPUART0SRC_MASK) | SIM_SOPT2_LPUART0SRC(3);
```
Shouldn't this be part of the `uart_init()` function and rather then be configurable from the `periph_conf.h` if needed?

gebart replied:
I agree, I will try to come up with a clean enhancement to the LPUART driver

gebart replied:
I have added a new define for the lpuart clock sources

haukepetersen commented on boards/frdm-kw41z/include/board.h:
```diff
@@ -0,0 +1,101 @@
+ */
+#define LED0_PIN            GPIO_PIN(PORT_B,  0)
+#define LED0_MASK           (1 << 0)
+#define LED0_ON            (GPIOB->PCOR = LED0_MASK)
```
something is off with the indention here.

kYc0o replied:
@gebart maybe our browsers show something different, can you check if this define is aligned?

gebart replied:
Addressed

haukepetersen commented on boards/frdm-kw41z/include/gpio_params.h:
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2017 SKF AB
```
Did you write this file for a different company?

gebart replied:
Yes, it was part of a project for a client

haukepetersen commented on boards/frdm-kw41z/include/periph_conf.h:
```diff
@@ -0,0 +1,295 @@
+};
+#define UART_NUMOF          (sizeof(uart_config) / sizeof(uart_config[0]))
+#define LPUART_0_ISR        isr_lpuart0
+
```
remove newline

gebart replied:
addressed

gebart commented:
@haukepetersen regarding the CPUs added, I added all three variants of the KW41Z because the 21z and 31z are subsets of the 41z and the headers are packaged together from NXP. If you think it is better to not pollute the tree we can remove them and re-add them when there is a use case, though most development boards will probably have the 41z variant for flexibility and only actual products will use the 21z for cost savings. We have one product board (not commercially available) using the 21z, but since it is out of the public master repo it does not matter where the vendor header ends up in that source tree.

emmanuelsearch commented:
I tried to compile on OS X and I get this lining error. Can you reproduce?

```
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /RIOT/sys/uart_stdio
kinetis.ld:30: nonconstant expression for origin
collect2: error: ld returned 1 exit status
make: *** [link] Error 1
```

emmanuelsearch commented:
Nevertheless, I could flash from Linux (via Vagrant) and I could verify that HelloWorld, SAUL and Javascript examples work fine ;)

haukepetersen commented:
@gebart regarding the headers: I see the connection and looking at this from that perspective I say lets keep them in. Please address the remaining comments and I think we are ready to merge this then.

kYc0o commented:
> I tried to compile on OS X and I get this lining error. Can you reproduce?
```
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /RIOT/sys/uart_stdio
kinetis.ld:30: nonconstant expression for origin
collect2: error: ld returned 1 exit status
make: *** [link] Error 1
```

In my OS X I can compile it without any issue. It's maybe the arm-none-eabi-gcc version which might be outdated...

kYc0o commented:
I have tested the SAUL interface and it seems that some of the ADC measurements are erroneous or I can't give any meaning, especially those with negative values. @gebart can you confirm that is normal?

emmanuelsearch commented:
never mind for the linking error: @kYc0o is right, an updated arm-none-eabi-gcc version solves the issue!

gebart commented:
@kYc0o RE: negative ADC readings:
SAUL uses int16_t for its values, but the ADC is using single ended measurements which give an unsigned 16 bit number, so any value > half the reference voltage will be seen as negative in the `saul read` shell command output. The exception is the ADC0_DP/ADC0_DM differential pair which is connected via J35 to an on board thermistor RT2, that channel is configured as a differential reading, so negative values are actual negative voltages across the thermistor (should not happen, since there are pull resistors on both sides)

I found an issue with the ADC in the current configuration. The internal VREF analog voltage reference module is not enabled, but the ADC periph driver for Kinetis is configured to always use that analog reference source. A better default for this CPU would be to use VDDA for analog reference by default, but let the user/board configuration override and turn on the VREF module if needed. The VREF should be off by default since it consumes some power, but using it may improve the ADC reading for small signals.
The other Kinetis CPUs currently in tree have external VREF pins which are used if the internal VREF module is off which is why this has not been an issue before, but the KW41Z seem to not have any connection when the VREF is off, and the readings look like garbage.
I need to fix this before we can merge this CPU.

haukepetersen commented:
Just did a quick test and flashed the board - it seems like something is wrong with the UART here, though I can't quickly say if this is related to the port or to my host. Output of the default example looks like this:
```
> help
2017-11-29 11:39:10,976 - INFO #  help
2017-11-29 11:39:10,981 - INFO # Command              Description
2017-11-29 11:39:10,982 - INFO # ---------------------------------------
2017-11-29 11:39:10,986 - INFO # reboot               Reboot the node
2017-11-29 11:39:10,991 - INFO # p.�ɥ��information about running threads.
2017-11-29 11:39:10,991 - INFO # rtc                  control RTC pY�
                                                                     ��Ʌ��interface
2017-11-29 11:39:11,001 - INFO # saul                 interact with sensors and actuators using SAUL
> ifconfig
2017-11-29 11:39:12,346 - INFO #  ifconfig
2017-11-29 11:39:12,346 - INFO # shell: command not found: ifconfig
> help
2017-11-29 11:39:13,656 - INFO #  helpC����+��             Description
2017-11-29 11:39:13,661 - INFO # ---------------------------------------
2017-11-29 11:39:13,666 - INFO # reboot               Reboot the node
2017-11-29 11:39:13,671 - INFO # ps                   Prints information about running threads.
2017-11-29 11:39:13,676 - INFO # rtc                  control RTC peripheral interface
2017-11-29 11:39:13,681 - INFO # saul                 interact with sensors and actuators using SAUL
> reboot
2017-11-29 11:39:15,628 - INFO #  reboot�main(): This is RIOT! (Version: 2018.01-devel-550-g20781-𮝵HEAD)
2017-11-29 11:39:15,628 - INFO # Welcome to RIOT!
> help
2017-11-29 11:39:16,973 - INFO #  help
2017-11-29 11:39:16,974 - INFO # Command              Description
2017-11-29 11:39:16,978 - INFO # ---------------------------------------
2017-11-29 11:39:16,983 - INFO # reboot               Reboot the node
2017-11-29 11:39:16,988 - INFO # ps                   Prints information about running threads.
2017-11-29 11:39:16,988 - INFO # rtc                  control RTC �Y�
                                                                     ��Ʌ�interface
2017-11-29 11:39:16,998 - INFO # saul                 interact with sensors and actuators using SAUL
> help
2017-11-29 11:39:18,338 - INFO #  help
2017-11-29 11:39:18,339 - INFO # Command              Description
2017-11-29 11:39:18,343 - INFO # ---------------------------------------
2017-11-29 11:39:18,348 - INFO # reboot               Reboot the node
2017-11-29 11:39:18,353 - INFO # ps                   Prints information about running threads.
2017-11-29 11:39:18,353 - INFO # rtc                  control RTC perip�Ʌ�interface
2017-11-29 11:39:18,363 - INFO # saul                 interact with sensors and actuators using SAUL
> help
2017-11-29 11:39:19,286 - INFO #  helpC����+���             DescripW��K���������������������������������������IH�����              Reboot the node
2017-11-29 11:39:19,291 - INFO # ps                   Prints information about running threads.
2017-11-29 11:39:19,296 - INFO # rtc                  control RTC pY�
                                                                     ��Ʌ�interface
2017-11-29 11:39:19,301 - INFO # saul                 interact with sensors and actuators using SAUL
> help
2017-11-29 11:39:20,626 - INFO #  hel�C����+��             DescripW��K���������������������������������������IH����с              Reboot the node
2017-11-29 11:39:20,631 - INFO # p.�ɥ��information about running threads.
2017-11-29 11:39:20,632 - INFO # rtc                  control RTC peri�Ʌ�interface
2017-11-29 11:39:20,641 - INFO # saul                 interact with sensors and actuators using SA
```
so random garbage every now and then. I have not seen this with any other board, so I tend to think this has something to do with the board. It is also not connected to `pyterm`, as using `cutecom` yields a similar output:
```
> help
Command              Description
---------------------------------------
reboot               Reboot the node
ps                   Prints information about running threads.
rtc                  control RTC \0xf0Y\0xae\0x0b\0xa1\0x95\0xc9\0x85\0xb1\0x02interface
saul                 interact with sensors and actuators using SAUL
> hel\0xf0C\0xa1\0xa8\0xab\0xab+\0xb9\0x91\0x02             DescripW\0xeb\0xcbK\0xa8\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9\0xa9IH\0x95\0x89\0xbd\0xbd\0xd1\0x02              Reboot the node
\0xf0.\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x82\0xc9\0xa5\0xb9\0xd1\0xcd\0x81information about running threads.
rtc                  control RTC \0xf0Y\0xae\0x0b\0xa1\0x95\0xc9\0x85\0xb1\0x02interface
saul                 interact with sensors and actuators using SAUL
> help
Command              Description
---------------------------------------
reboot               Reboot the node
p.\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x02\0x82\0xc9\0xa5\0xb9\0xd1\0xcd\0x02information about running threads.
rtc                  control RTC peripheral interface
saul                 interact with sensors and actuators using SAUL
> 
```

gebart commented:
@haukepetersen I'm guessing your UART issues is related to the clocking. The current configuration uses the internal RC oscillator in the CPU for the core clock, this is less accurate than using the onboard crystal, but it uses less power. Could you try again with the latest version of this PR branch?

As for the ADC, I will add a bare minimum vref driver to the PR so that the ADC gives reasonable values for this board

gebart commented:
rebased on top of #8560

kYc0o commented on boards/frdm-kw41z/board.c:
```diff
@@ -0,0 +1,51 @@
+#if MODULE_XTIMER
+    /* Start the RTT, used as time base for xtimer */
+    rtt_init();
+#endif
```
RTT cannot be selected independently of xtimer needs? Besides, this is only required for this board? I tend to think that this is more a CPU requirement, but I can be mistaken.

gebart replied:
It is because this board uses the lptmr hardware timer for xtimer, while the other boards use the PIT hardware timer for xtimer. the LPTMR driver implementation uses RTT for a time base reference. I would like to improve the lptmr driver, but limitations in the hardware in combination with requirements from the xtimer implementation makes it difficult to make a good implementation without a separate time reference right now.
If you like I can change the default config to use PIT like the other kinetis boards

kYc0o commented on boards/frdm-kw41z/include/periph_conf.h:
```diff
@@ -0,0 +1,307 @@
+    .enable_oscillator = true, /* Use RF module oscillator */
+    .select_fast_irc = true,
+    .enable_mcgirclk = true, /* Used for LPUART clocking */
+};
```
Can the values of this struct be aligned for readability?

gebart replied:
addressed

kYc0o commented on cpu/kinetis/periph/pwm.c:
```diff
@@ -29,6 +29,8 @@
 #include "assert.h"
 #include "periph/pwm.h"
 
+#ifdef FTM0
```
Can you add some comments on what this means? It looks strange that before it worked without this `#ifdef` and now we need it...

kYc0o commented:
Tested again without trouble, just found some small things.

gebart commented on cpu/kinetis/periph/gpio.c:
```diff
@@ -347,3 +352,12 @@ void isr_portg(void)
+
+#if defined(PORTB_BASE) && defined(PORTC_BASE)
+/* Combined ISR used in certain KL devices */
+void isr_portbc(void)
```
the name was wrong, thanks for noticing!

kYc0o commented:
It looks ok for me. @gebart did you addressed @haukepetersen concerns?

gebart commented:
Murdock is unhappy, will post back when I have addressed the problems.

gebart commented:
Fixed the build errors that Murdock was complaining about.

kYc0o commented:
Well, Murdock is still complaining, doxygen and commit messages (but most are squash).

gebart commented:
@kYc0o fixed all Murdock errors except for the needs squashing check

kYc0o commented:
There's only a small thing left:

```
cpu/kinetis/include/bme.h: no / broken header guard
```

You may squash everything and amend this last fix, so we can have an all green!

gebart commented:
@haukepetersen do you have anything to add here or can we dismiss your review?

haukepetersen commented:
Sorry for being unresponsive, not looking into github too much recently... I don't have time to review this PR again, but everything seems in order. I simply dismiss my change requests and so you guys are free to proceed. Sorry again for blocking this!

kYc0o commented:
Everything looks green here. Maybe @emmanuelsearch wants to make a final test? I tested everything but my board has OpenSDA firmware for flashing, maybe a test on a board with JLink is needed?

emmanuelsearch commented:
I'm able to flash and run examples on the board from my Linux VM with J-Link OpenSDA.
But on OS X I'm unable to flash, I get:

```
Info : No device selected, using first device.
Error: No J-Link device found.

make: *** [flash] Error 1
```

Any advice on how to change that?

kYc0o commented:
JLink and OpenSDA are two different firmwares present on the flasher, which one is on your board?

emmanuelsearch commented:
no idea, how should I check that?

kYc0o commented:
Can you paste the successful flashing output of your Linux VM?

emmanuelsearch commented:
```
### Flashing Target ###
/home/vagrant/RIOT/examples/saul/bin/frdm-kw41z/saul_example.elf is fine.
Open On-Chip Debugger 0.10.0+dev-00167-g29cfe9c (2017-07-13-13:14)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : add flash_bank kinetis klx.pflash
adapter speed: 1000 kHz
none separate
cortex_m reset_config sysresetreq
srst_only separate srst_nogate srst_open_drain connect_deassert_srst
Info : No device selected, using first device.
Info : J-Link OpenSDA 2 compiled May  6 2016 11:04:17
Info : Hardware version: 1.00
Info : VTarget = 3.300 V
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x0bc11477
Info : MDM: Chip is unsecured. Continuing.
Info : klx.cpu: hardware has 2 breakpoints, 2 watchpoints
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* klx.cpu            cortex_m   little klx.cpu            running
Info : MDM: Chip is unsecured. Continuing.
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x0000075c msp: 0x1fff8200
auto erase enabled
Info : Kinetis MKW41Z512xxx4 detected: 2 flash blocks
Info : 2 PFlash banks: 512k total
Info : This device supports Program Longword execution only.
Info : Disabling Kinetis watchdog (initial SIM_COPC 0x0c)
Info : This device supports Program Longword execution only.
Warn : Flash Configuration Field written.
Warn : Reset or power off the device to make settings effective.
Info : This device supports Program Longword execution only.
wrote 16384 bytes from file /home/vagrant/RIOT/examples/saul/bin/frdm-kw41z/saul_example.elf in 0.641436s (24.944 KiB/s)
verified 16188 bytes in 0.216723s (72.944 KiB/s)
Info : MDM: Chip is unsecured. Continuing.
shutdown command invoked
Done flashing
```

kYc0o commented:
Oh I see, yeah, I was confused, my testings were done with a DAP firmware. Maybe your openocd version on OS X is not the latest? Can you verify?

emmanuelsearch commented:
on OS X I have:

```
Open On-Chip Debugger 0.10.0+dev-00167-g29cfe9c5 (2017-07-13-13:24)
```

kYc0o commented:
Ok then I think it's related to the USB being captured by the VM, are you sure the VM is completely off when you test on OS X? Or that your device is not being captured by the VM?

emmanuelsearch commented:
My VM is completely off. Will try to enquire further.
Are you able to flash from OS X ?

kYc0o commented:
Yes I'm able to flash but as I pointed out in a previous comment, I have flashed the DAP firmware to my board, thus I need to to the following:

```console
DEBUG_ADAPTER=dap BOARD=frdm-kw41z make -C examples/default flash term
```

which results in a successful flashing.

gebart commented:
@kYc0o @haukepetersen @emmanuelsearch Thanks for testing

kYc0o commented:
I just grabbed @emmanuelsearch board with JLink firmware and successfully flashed. Thus I suspect there's something in the configuration of either openocd or the usb ports on @emmanuelsearch setup.

kYc0o commented:
I'm about to merge this PR but still have a small doubt. @gebart can you check if the modification to the existing files are completely coupled with this PR goals? It's just that I see some commits which modify some behaviour (e.g. 6c7c3e1a9d78a06c3651084ef8881341e1499018 ) which looks "somehow" unrelated. I'm not against merging this PR as is, but just for consistency I'd prefer to keep only the commits for this PR and if needed separate the rest in other PRs (which can be merged asap given that they were already reviewed and tested here).

gebart commented:
Will split some commits

kYc0o commented:
Wow excellent!! Congrats @gebart this was a long trip!

emmanuelsearch commented:
@gebart great, very much looking forward to #7107 ;)

gebart commented:
Thank you for the help with reviewing this pretty big PR!
